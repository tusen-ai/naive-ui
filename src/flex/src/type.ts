import { type Property } from 'csstype'

export type FlexAlign = Property.AlignItems

export type FlexJustify = Property.JustifyContent
