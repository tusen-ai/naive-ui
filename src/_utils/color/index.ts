import { composite } from 'seemly'

export * from './resolve-color'

export function createHoverColor (rgb: string): string {
  return composite(rgb, [255, 255, 255, 0.16])
}

export function createPressedColor (rgb: string): string {
  return composite(rgb, [0, 0, 0, 0.12])
}
