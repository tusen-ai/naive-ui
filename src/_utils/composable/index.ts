export { useAdjustedTo } from './use-adjusted-to'
export {
  useInjectionCollection,
  useInjectionElementCollection,
  useInjectionInstanceCollection
} from './use-collection'
export { useDeferredTrue } from './use-deferred-true'
export { useHoudini } from './use-houdini'
export { useIsComposing } from './use-is-composing'
export {
  lockHtmlScrollRightCompensationRef,
  useLockHtmlScroll
} from './use-lock-html-scroll'
export { useReactivated } from './use-reactivated'
export { useOnResize } from './use-resize'
