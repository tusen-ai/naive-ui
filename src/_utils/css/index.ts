export { formatLength } from './format-length'
export { color2Class } from './color-to-class'
export { rtlInset } from './rtl-inset'
