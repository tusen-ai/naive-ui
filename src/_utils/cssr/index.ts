/* eslint-disable @typescript-eslint/restrict-template-expressions */
import { CssRender, type CNode, type CProperties } from 'css-render'
import { plugin as BemPlugin } from '@css-render/plugin-bem'

const namespace = 'n'
const prefix = `.${namespace}-`
const elementPrefix = '__'
const modifierPrefix = '--'

const cssr = CssRender()
const plugin = BemPlugin({
  blockPrefix: prefix,
  elementPrefix,
  modifierPrefix
})
cssr.use(plugin)
const { c, find } = cssr
const { cB, cE, cM, cNotM } = plugin

function insideModal (style: CNode): CNode {
  return c(
    ({ props: { bPrefix } }) =>
      `${bPrefix || prefix}modal, ${bPrefix || prefix}drawer`,
    [style]
  )
}

function insidePopover (style: CNode): CNode {
  return c(({ props: { bPrefix } }) => `${bPrefix || prefix}popover`, [style])
}

function asModal (style: CProperties): CNode {
  return c(({ props: { bPrefix } }) => `&${bPrefix || prefix}modal`, style)
}

// child block
const cCB: typeof cB = ((...args: any[]) => {
  return c('>', [(cB as any)(...args)])
}) as any

export {
  c,
  cB,
  cE,
  cM,
  cNotM,
  cCB,
  insideModal,
  insidePopover,
  asModal,
  prefix,
  namespace,
  find
}

export { createKey } from './create-key'
