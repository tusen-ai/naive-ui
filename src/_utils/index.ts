export {
  call,
  keep,
  omit,
  flatten,
  getSlot,
  getVNodeChildren,
  keysOf,
  render,
  getFirstSlotVNode,
  createDataKey,
  createRefSetter,
  createInjectionKey,
  resolveSlot,
  resolveSlotWithProps,
  resolveWrappedSlot,
  isSlotEmpty,
  mergeEventHandlers,
  isNodeVShowFalse,
  resolveWrappedSlotWithProps,
  Wrapper
} from './vue'
export type { MaybeArray } from './vue'
export {
  warn,
  warnOnce,
  throwError,
  smallerSize,
  largerSize,
  getTitleAttribute
} from './naive'
export type {
  ExtractPublicPropTypes,
  ExtractInternalPropTypes,
  Mutable
} from './naive'
export { formatLength, color2Class } from './css'
export { createKey } from './cssr'
export { isJsdom } from './env/is-jsdom'
export { isBrowser } from './env/is-browser'
export { eventEffectNotPerformed, markEventEffectPerformed } from './event'
export * from './composable'
export * from './dom'
