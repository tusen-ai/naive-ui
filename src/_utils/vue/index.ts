export { getSlot } from './get-slot'
export { getVNodeChildren } from './get-v-node-children'
export { keep } from './keep'
export { omit } from './omit'
export { flatten } from './flatten'
export { call } from './call'
export { keysOf } from './keysOf'
export { render } from './render'
export { getFirstSlotVNode } from './get-first-slot-vnode'
export { createDataKey } from './create-data-key'
export { createRefSetter } from './create-ref-setter'
export { createInjectionKey } from './create-injection-key'
export {
  resolveSlot,
  resolveWrappedSlot,
  resolveSlotWithProps,
  resolveWrappedSlotWithProps,
  isSlotEmpty
} from './resolve-slot'
export type { MaybeArray } from './call'
export { mergeEventHandlers } from './merge-handlers'
export { isNodeVShowFalse } from './is-node-v-show-false'
export { Wrapper } from './wrapper'
