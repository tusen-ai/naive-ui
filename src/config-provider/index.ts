export {
  configProviderProps,
  default as NConfigProvider
} from './src/ConfigProvider'
export type { ConfigProviderProps } from './src/ConfigProvider'
export * from './src/interface'
