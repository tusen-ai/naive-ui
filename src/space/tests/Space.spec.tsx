import { mount, type VueWrapper } from '@vue/test-utils'
import { h, Fragment, createCommentVNode } from 'vue'
import { NSpace } from '../index'
import { NConfigProvider } from '../../config-provider'
import type { Justify } from '../src/Space'
import { c } from '../../_utils/cssr'

const getChildrenNode = (wrapper: VueWrapper<any>): any[] => {
  return (
    wrapper.findAll('div').filter((v) => {
      return !v.classes().includes('n-space')
    }) || []
  )
}

describe('n-space', () => {
  it('should work with import on demand', () => {
    mount(NSpace)
  })

  it('render empty children', () => {
    const wrapper = mount({
      render () {
        return <NSpace />
      }
    })
    expect(wrapper.find('.n-space')).not.toBe(null)
    expect(wrapper.html()).toMatchSnapshot()
    wrapper.unmount()
  })

  it('render space string size', () => {
    const wrapper = mount({
      render () {
        return <NSpace size="large">{{ default: () => 'kirby' }}</NSpace>
      }
    })
    expect(wrapper.attributes('style')).toContain('margin')
    expect(wrapper.html()).toMatchSnapshot()
    wrapper.unmount()
  })

  it('render space number size', () => {
    const size = 20
    const wrapper = mount({
      render () {
        return <NSpace size={size}>{{ default: () => 'kirby' }}</NSpace>
      }
    })
    expect(wrapper.attributes('style')).toContain(`margin-top: -${size / 2}px`)
    expect(wrapper.html()).toMatchSnapshot()
    wrapper.unmount()
  })

  it('render space array size', async () => {
    const wrapper = mount({
      render () {
        return (
          <NSpace size={[20, 30]}>
            {{
              default: () => [<div>1</div>, <div>2</div>]
            }}
          </NSpace>
        )
      }
    })

    const childNodes = getChildrenNode(wrapper)
    expect(childNodes[0].attributes('style')).toContain('margin-right: 20px;')

    await wrapper.setProps({ vertical: true })
    expect(childNodes[0].attributes('style')).toContain('margin-bottom: 30px;')
    wrapper.unmount()
  })

  it('render vertical space', () => {
    const wrapper = mount({
      render () {
        return (
          <NSpace vertical>
            {{
              default: () => [<div>1</div>, <div>2</div>]
            }}
          </NSpace>
        )
      }
    })
    expect(wrapper.attributes('style')).toContain('flex-direction: column;')
    expect(wrapper.html()).toMatchSnapshot()
    wrapper.unmount()
  })

  it('should work with `inline` prop', () => {
    const wrapper = mount(NSpace, {
      props: {
        inline: true
      },
      slots: {
        default: () => '07akioni'
      }
    })

    expect(wrapper.attributes('style')).toContain('display: inline-flex;')
    wrapper.unmount()
  })

  it('should render with invalidElement', () => {
    const wrapper = mount({
      render () {
        return (
          <NSpace>
            {{
              default: () => (
                <>
                  text1<span>text1</span>
                  text1
                </>
              )
            }}
          </NSpace>
        )
      }
    })

    const childNodes = getChildrenNode(wrapper)
    expect(childNodes.length).toBe(3)
    expect(wrapper.html()).toMatchSnapshot()
    wrapper.unmount()
  })

  it('render justify space', () => {
    const justifyList: Justify[] = [
      'start',
      'end',
      'center',
      'space-around',
      'space-between',
      'space-evenly'
    ]

    justifyList.forEach((pos) => {
      const wrapper = mount({
        render () {
          return (
            <NSpace justify={pos}>
              {{
                default: () => [<div>1</div>, <div>2</div>]
              }}
            </NSpace>
          )
        }
      })

      expect(wrapper.attributes('style')).toContain(
        `justify-content: ${
          ['start', 'end'].includes(pos) ? 'flex-' + pos + ';' : pos
        }`
      )
      expect(wrapper.html()).toMatchSnapshot()
      wrapper.unmount()
    })

    justifyList.forEach((pos) => {
      const wrapper = mount({
        render () {
          return (
            <NConfigProvider rtl={[{ name: 'Space', style: c(null) }]}>
              {{
                default: () => {
                  return (
                    <NSpace justify={pos}>
                      {{
                        default: () => [<div>1</div>, <div>2</div>]
                      }}
                    </NSpace>
                  )
                }
              }}
            </NConfigProvider>
          )
        }
      })

      expect(wrapper.find('.n-space').attributes('style')).toContain(
        `justify-content: ${
          ['start', 'end'].includes(pos) ? 'flex-' + pos + ';' : pos
        }`
      )
      expect(wrapper.html()).toMatchSnapshot()
      wrapper.unmount()
    })
  })

  it('render custom style space', () => {
    const wrapper = mount(
      {
        render () {
          return <NSpace>{{ default: () => 'div' }}</NSpace>
        }
      },
      {
        props: {
          itemStyle: { backgroundColor: 'red', color: 'yellow' }
        }
      }
    )
    const childNodes = getChildrenNode(wrapper)

    expect(childNodes[0].attributes('style')).toContain(
      'background-color: red; color: yellow;'
    )
    expect(wrapper.html()).toMatchSnapshot()
    wrapper.unmount()
  })

  it('should not render while v-if is false', () => {
    const wrapper = mount({
      render () {
        return <NSpace>{{ default: () => false && 'div' }}</NSpace>
      }
    })
    const childNodes = getChildrenNode(wrapper)
    expect(childNodes.length).toEqual(0)
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should work with `wrap` prop', async () => {
    const wrapper = mount({
      render () {
        return (
          <NSpace>{{ default: () => [<div>1</div>, <div>2</div>] }}</NSpace>
        )
      }
    })
    expect(wrapper.find('.n-space').attributes('style')).toContain(
      'flex-wrap: wrap'
    )

    await wrapper.setProps({ wrap: false })
    expect(wrapper.find('.n-space').attributes('style')).toContain(
      'flex-wrap: nowrap'
    )
    wrapper.unmount()
  })

  it('should not render while slot is Comment', () => {
    const wrapper = mount({
      render () {
        return (
          <NSpace>
            {{
              default: () => createCommentVNode('random comment text')
            }}
          </NSpace>
        )
      }
    })
    const childNodes = getChildrenNode(wrapper)
    expect(childNodes.length).toEqual(0)
    expect(wrapper.html()).toMatchSnapshot()
    wrapper.unmount()
  })

  it('shound not render a container to wrap child elements', () => {
    const wrapper = mount(
      {
        render () {
          return (
            <NSpace>
              {{
                default: () => [<div>1</div>, <div>2</div>]
              }}
            </NSpace>
          )
        }
      },
      {
        props: {
          wrapItem: false,
          internalUseGap: true
        }
      }
    )

    expect(wrapper.html()).toMatchSnapshot()
    wrapper.unmount()
  })
})
