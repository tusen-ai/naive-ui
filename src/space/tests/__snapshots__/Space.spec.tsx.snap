// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`n-space render custom style space 1`] = `
"<div role="none" class="n-space" style="display: flex; flex-direction: row; justify-content: flex-start; flex-wrap: wrap; margin-top: -4px; margin-bottom: -4px;">
  <div role="none" style="background-color: red; color: yellow; max-width: 100%; padding-top: 4px; padding-bottom: 4px;">div</div>
</div>"
`;

exports[`n-space render empty children 1`] = `""`;

exports[`n-space render justify space 1`] = `
"<div role="none" class="n-space" style="display: flex; flex-direction: row; justify-content: flex-start; flex-wrap: wrap; margin-top: -4px; margin-bottom: -4px;">
  <div role="none" style="max-width: 100%; margin-right: 12px; padding-top: 4px; padding-bottom: 4px;">
    <div>1</div>
  </div>
  <div role="none" style="max-width: 100%; padding-top: 4px; padding-bottom: 4px;">
    <div>2</div>
  </div>
</div>"
`;

exports[`n-space render justify space 2`] = `
"<div role="none" class="n-space" style="display: flex; flex-direction: row; justify-content: flex-end; flex-wrap: wrap; margin-top: -4px; margin-bottom: -4px;">
  <div role="none" style="max-width: 100%; margin-right: 12px; padding-top: 4px; padding-bottom: 4px;">
    <div>1</div>
  </div>
  <div role="none" style="max-width: 100%; padding-top: 4px; padding-bottom: 4px;">
    <div>2</div>
  </div>
</div>"
`;

exports[`n-space render justify space 3`] = `
"<div role="none" class="n-space" style="display: flex; flex-direction: row; justify-content: center; flex-wrap: wrap; margin-top: -4px; margin-bottom: -4px;">
  <div role="none" style="max-width: 100%; margin-right: 12px; padding-top: 4px; padding-bottom: 4px;">
    <div>1</div>
  </div>
  <div role="none" style="max-width: 100%; padding-top: 4px; padding-bottom: 4px;">
    <div>2</div>
  </div>
</div>"
`;

exports[`n-space render justify space 4`] = `
"<div role="none" class="n-space" style="display: flex; flex-direction: row; justify-content: space-around; flex-wrap: wrap; margin-top: -4px; margin-bottom: -4px;">
  <div role="none" style="max-width: 100%; margin-right: 6px; margin-left: 6px; padding-top: 4px; padding-bottom: 4px;">
    <div>1</div>
  </div>
  <div role="none" style="max-width: 100%; margin-right: 6px; margin-left: 6px; padding-top: 4px; padding-bottom: 4px;">
    <div>2</div>
  </div>
</div>"
`;

exports[`n-space render justify space 5`] = `
"<div role="none" class="n-space" style="display: flex; flex-direction: row; justify-content: space-between; flex-wrap: wrap; margin-top: -4px; margin-bottom: -4px;">
  <div role="none" style="max-width: 100%; margin-right: 6px; padding-top: 4px; padding-bottom: 4px;">
    <div>1</div>
  </div>
  <div role="none" style="max-width: 100%; margin-left: 6px; padding-top: 4px; padding-bottom: 4px;">
    <div>2</div>
  </div>
</div>"
`;

exports[`n-space render justify space 6`] = `
"<div role="none" class="n-space" style="display: flex; flex-direction: row; justify-content: space-evenly; flex-wrap: wrap; margin-top: -4px; margin-bottom: -4px;">
  <div role="none" style="max-width: 100%; margin-right: 6px; margin-left: 6px; padding-top: 4px; padding-bottom: 4px;">
    <div>1</div>
  </div>
  <div role="none" style="max-width: 100%; margin-right: 6px; margin-left: 6px; padding-top: 4px; padding-bottom: 4px;">
    <div>2</div>
  </div>
</div>"
`;

exports[`n-space render justify space 7`] = `
"<div class="n-config-provider">
  <div role="none" class="n-space n-space--rtl" style="display: flex; flex-direction: row; justify-content: flex-start; flex-wrap: wrap; margin-top: -4px; margin-bottom: -4px;">
    <div role="none" style="max-width: 100%; margin-left: 12px; padding-top: 4px; padding-bottom: 4px;">
      <div>1</div>
    </div>
    <div role="none" style="max-width: 100%; padding-top: 4px; padding-bottom: 4px;">
      <div>2</div>
    </div>
  </div>
</div>"
`;

exports[`n-space render justify space 8`] = `
"<div class="n-config-provider">
  <div role="none" class="n-space n-space--rtl" style="display: flex; flex-direction: row; justify-content: flex-end; flex-wrap: wrap; margin-top: -4px; margin-bottom: -4px;">
    <div role="none" style="max-width: 100%; margin-left: 12px; padding-top: 4px; padding-bottom: 4px;">
      <div>1</div>
    </div>
    <div role="none" style="max-width: 100%; padding-top: 4px; padding-bottom: 4px;">
      <div>2</div>
    </div>
  </div>
</div>"
`;

exports[`n-space render justify space 9`] = `
"<div class="n-config-provider">
  <div role="none" class="n-space n-space--rtl" style="display: flex; flex-direction: row; justify-content: center; flex-wrap: wrap; margin-top: -4px; margin-bottom: -4px;">
    <div role="none" style="max-width: 100%; margin-left: 12px; padding-top: 4px; padding-bottom: 4px;">
      <div>1</div>
    </div>
    <div role="none" style="max-width: 100%; padding-top: 4px; padding-bottom: 4px;">
      <div>2</div>
    </div>
  </div>
</div>"
`;

exports[`n-space render justify space 10`] = `
"<div class="n-config-provider">
  <div role="none" class="n-space n-space--rtl" style="display: flex; flex-direction: row; justify-content: space-around; flex-wrap: wrap; margin-top: -4px; margin-bottom: -4px;">
    <div role="none" style="max-width: 100%; margin-left: 6px; margin-right: 6px; padding-top: 4px; padding-bottom: 4px;">
      <div>1</div>
    </div>
    <div role="none" style="max-width: 100%; margin-left: 6px; margin-right: 6px; padding-top: 4px; padding-bottom: 4px;">
      <div>2</div>
    </div>
  </div>
</div>"
`;

exports[`n-space render justify space 11`] = `
"<div class="n-config-provider">
  <div role="none" class="n-space n-space--rtl" style="display: flex; flex-direction: row; justify-content: space-between; flex-wrap: wrap; margin-top: -4px; margin-bottom: -4px;">
    <div role="none" style="max-width: 100%; margin-left: 6px; padding-top: 4px; padding-bottom: 4px;">
      <div>1</div>
    </div>
    <div role="none" style="max-width: 100%; margin-right: 6px; padding-top: 4px; padding-bottom: 4px;">
      <div>2</div>
    </div>
  </div>
</div>"
`;

exports[`n-space render justify space 12`] = `
"<div class="n-config-provider">
  <div role="none" class="n-space n-space--rtl" style="display: flex; flex-direction: row; justify-content: space-evenly; flex-wrap: wrap; margin-top: -4px; margin-bottom: -4px;">
    <div role="none" style="max-width: 100%; margin-left: 6px; margin-right: 6px; padding-top: 4px; padding-bottom: 4px;">
      <div>1</div>
    </div>
    <div role="none" style="max-width: 100%; margin-left: 6px; margin-right: 6px; padding-top: 4px; padding-bottom: 4px;">
      <div>2</div>
    </div>
  </div>
</div>"
`;

exports[`n-space render space number size 1`] = `
"<div role="none" class="n-space" style="display: flex; flex-direction: row; justify-content: flex-start; flex-wrap: wrap; margin-top: -10px; margin-bottom: -10px;">
  <div role="none" style="max-width: 100%; padding-top: 10px; padding-bottom: 10px;">kirby</div>
</div>"
`;

exports[`n-space render space string size 1`] = `
"<div role="none" class="n-space" style="display: flex; flex-direction: row; justify-content: flex-start; flex-wrap: wrap; margin-top: -6px; margin-bottom: -6px;">
  <div role="none" style="max-width: 100%; padding-top: 6px; padding-bottom: 6px;">kirby</div>
</div>"
`;

exports[`n-space render vertical space 1`] = `
"<div role="none" class="n-space" style="display: flex; flex-direction: column; justify-content: flex-start; flex-wrap: nowrap;">
  <div role="none" style="max-width: 100%; margin-bottom: 8px;">
    <div>1</div>
  </div>
  <div role="none" style="max-width: 100%;">
    <div>2</div>
  </div>
</div>"
`;

exports[`n-space should not render while slot is Comment 1`] = `""`;

exports[`n-space should not render while v-if is false 1`] = `""`;

exports[`n-space should render with invalidElement 1`] = `
"<div role="none" class="n-space" style="display: flex; flex-direction: row; justify-content: flex-start; flex-wrap: wrap; margin-top: -4px; margin-bottom: -4px;">
  <div role="none" style="max-width: 100%; margin-right: 12px; padding-top: 4px; padding-bottom: 4px;">text1</div>
  <div role="none" style="max-width: 100%; margin-right: 12px; padding-top: 4px; padding-bottom: 4px;"><span>text1</span></div>
  <div role="none" style="max-width: 100%; padding-top: 4px; padding-bottom: 4px;">text1</div>
</div>"
`;

exports[`n-space shound not render a container to wrap child elements 1`] = `
"<div role="none" class="n-space" style="display: flex; flex-direction: row; justify-content: flex-start; flex-wrap: wrap; margin-top: -4px; margin-bottom: -4px;">
  <div>1</div>
  <div>2</div>
</div>"
`;
